core/trino-main/src/main/java/io/trino/operator/output/PagePartitioner.java

```diff
@@ -153,6 +153,27 @@ public void partitionPage(Page page, OperatorContext operatorContext)
         operatorContext.recordOutput(positionsAppendersSizeAfter - positionsAppendersSizeBefore, outputPositionCount);
     }
 
+    /**
+     * Prepares this {@link PagePartitioner} for release to the pool by checking for dictionary mode appenders and either flattening
+     * them into direct appenders or forcing their current pages to flush to preserve a valuable dictionary encoded representation. This
+     * is done before release because we know that after reuse, the appenders will not observe any more inputs using the same dictionary.
+     */
+    public void prepareForRelease(OperatorContext operatorContext)
+    {
+        long positionsAppendersSizeBefore = getPositionsAppendersSizeInBytes();
+        for (int partition = 0; partition < positionsAppenders.length; partition++) {
+            PositionsAppenderPageBuilder positionsAppender = positionsAppenders[partition];
+            if (positionsAppender.shouldFlushBeforeRelease()) {
+                enqueuePage(positionsAppender.build(), partition);
+            }
+        }
+        long positionsAppendersSizeAfter = getPositionsAppendersSizeInBytes();
+        updateMemoryUsage();
+
+        // output positions has already been accounted for, but output size has not
+        operatorContext.recordOutput(positionsAppendersSizeAfter - positionsAppendersSizeBefore, 0);
```

Comment:
that seems wrong. Both should already be accounted for in `partitionPage`. Also flushing would make after size smaller than before. `partitionPage` specifically accounts before flushing because flushing can happen at later time.
PTAL @lukasz-stec

Reply:
I think you're right. This is already accounted for as part of the partitioning operation, sort of. In the case of dictionaries we know the results are _very_ incorrect because of the way that the appender size is calculated, so I guess not reporting the result of any dictionary to direct flattenings that might occur here doesn't really make the situation any worse than it already is.